File: backend/app/services/history_service.py
from typing import Dict, Any, List, Optional
from datetime import datetime, timedelta
from collections import defaultdict
import logging
from sqlalchemy.orm import Session
from app.core.config_loader import config_loader
from app.models.form import FormInstance, FormHistory, FormTemplate
from app.services.ontology_service import OntologyService

logger = logging.getLogger("history_service")


class HistoryService:
    
    @classmethod
    def get_recommend_values(
        cls, 
        form_code: str, 
        field_code: str, 
        user_id: str = None,
        db: Session = None,
        limit: int = None
    ) -> Dict[str, Any]:
        recommendations = []
        config = config_loader.get_recommendation_config()
        
        if limit is None:
            limit = config.get('recommendationLimit', 10)
        
        if db:
            db_recommendations = cls._get_db_recommendations(db, form_code, field_code, user_id, config)
            recommendations.extend(db_recommendations)
        
        static_recommendations = config_loader.get_recommendations(form_code, field_code)
        for rec in static_recommendations:
            if rec not in [r['value'] for r in recommendations]:
                recommendations.append({
                    'value': rec,
                    'score': 0.5,
                    'source': 'static'
                })
        
        sorted_recommendations = sorted(
            recommendations, 
            key=lambda x: x['score'], 
            reverse=True
        )
        
        # ⚠️ 重要：返回编码和标签，而不是仅返回值
        top_recommendations = []
        for r in sorted_recommendations[:limit]:
            code = r['value']  # 编码
            label = cls._get_field_label(form_code, field_code, code)  # 中文标签
            top_recommendations.append({
                'value': code,    # 用于提交
                'label': label,   # 用于显示
                'score': r['score'],
                'source': r.get('source', 'db')
            })
        
        return {
            "success": True,
            "recommendations": top_recommendations,
            "source": "hybrid" if db else "static"
        }
    
    @classmethod
    def _get_db_recommendations(
        cls, 
        db: Session, 
        form_code: str, 
        field_code: str, 
        user_id: Optional[str] = None,
        config: Dict = None
    ) -> List[Dict[str, Any]]:
        if config is None:
            config = config_loader.get_recommendation_config()
        
        value_scores = defaultdict(lambda: {'count': 0, 'last_used': None, 'user_count': 0})

        query_limit = config.get('historyQueryLimit', 1000)

        # [修复] 先通过 form_code 查出 template_id 列表，再按 template_id 过滤 FormInstance
        # 避免不同表单类型的字段值串入推荐结果
        template_ids = db.query(FormTemplate.id).filter(
            FormTemplate.form_code == form_code,
            FormTemplate.is_active == True
        ).all()
        tid_list = [t.id for t in template_ids]

        if not tid_list:
            # [兜底] FormTemplate 无记录时，直接从 form_history 表查询
            # 这确保即使未通过 upsert 创建模板，历史推荐仍能工作
            logger.info("[HistoryService] form_code=%s 无FormTemplate记录，回退到form_history表", form_code)
            return cls._get_recommendations_from_history(db, form_code, field_code, user_id, config)

        base_query = db.query(FormInstance).filter(
            FormInstance.template_id.in_(tid_list),
            FormInstance.status == 'submitted'
        )
        if user_id:
            base_query = base_query.filter(FormInstance.user_id == user_id)

        form_instances = base_query.order_by(
            FormInstance.submitted_at.desc()
        ).limit(query_limit).all()
        
        for instance in form_instances:
            form_data = instance.data or {}
            if field_code in form_data:
                value = str(form_data[field_code])
                if value and value.strip():
                    score_info = value_scores[value]
                    score_info['count'] += 1
                    
                    if instance.submitted_at:
                        if not score_info['last_used'] or instance.submitted_at > score_info['last_used']:
                            score_info['last_used'] = instance.submitted_at
                    
                    if user_id and instance.user_id == user_id:
                        score_info['user_count'] += 1
        
        recommendations = []
        now = datetime.now()
        
        for value, info in value_scores.items():
            score = cls._calculate_score(info, now, config)
            recommendations.append({
                'value': value,
                'score': score,
                'source': 'database',
                'count': info['count'],
                'user_count': info['user_count']
            })
        
        return sorted(recommendations, key=lambda x: x['score'], reverse=True)[:config.get('recommendationLimit', 10)]
    
    @classmethod
    def _calculate_score(cls, info: Dict[str, Any], now: datetime, config: Dict = None) -> float:
        if config is None:
            config = config_loader.get_recommendation_config()
        
        count = info['count']
        user_count = info['user_count']
        last_used = info['last_used']
        
        count_score_weight = config.get('countScoreWeight', 0.4)
        user_score_weight = config.get('userScoreWeight', 0.4)
        time_score_weight = config.get('timeScoreWeight', 0.2)
        count_score_per_unit = config.get('countScorePerUnit', 0.1)
        user_score_per_unit = config.get('userScorePerUnit', 0.2)
        time_decay_days = config.get('timeDecayDays', 30)
        
        count_score = min(count * count_score_per_unit, 1.0)
        
        user_score = min(user_count * user_score_per_unit, 1.0)
        
        time_score = 1.0
        if last_used:
            days_since = (now - last_used).days
            if days_since > 0:
                time_score = max(0.0, 1.0 - (days_since / time_decay_days))
        
        final_score = (
            count_score * count_score_weight + 
            user_score * user_score_weight + 
            time_score * time_score_weight
        )
        
        return final_score
    
    @classmethod
    def _get_field_label(cls, form_code: str, field_code: str, code: str) -> str:
        """
        从本体定义中获取字段的中文标签
        
        Args:
            form_code: 表单编码
            field_code: 字段编码
            code: 字段值（编码）
        
        Returns:
            中文标签，如果没有映射则返回原值
        """
        try:
            # 加载本体
            ontology_result = OntologyService.get_form_constraint(form_code)
            if not ontology_result.get('success'):
                return code
            
            ontology = ontology_result.get('constraints', {})
            
            # ⚠️ 本体结构是 entities 数组，需要遍历查找字段
            entities = ontology.get('entities', [])
            for entity in entities:
                fields = entity.get('fields', [])
                for field_def in fields:
                    if field_def.get('fieldCode') == field_code:
                        # 检查是否有 options 映射
                        options = field_def.get('options', [])
                        for opt in options:
                            # 支持两种格式：字符串数组 和 {value, label} 对象数组
                            if isinstance(opt, str):
                                if opt == code:
                                    return opt
                            elif isinstance(opt, dict):
                                if opt.get('value') == code:
                                    return opt.get('label', code)
                        
                        # 没有找到映射，返回原值
                        return code
            
            # 没有找到字段定义，返回原值
            return code
            
        except Exception as e:
            logger.warning("[HistoryService] 获取字段标签失败: %s", e)
            return code

    @classmethod
    def _get_recommendations_from_history(
        cls,
        db: Session,
        form_code: str,
        field_code: str,
        user_id: Optional[str] = None,
        config: Dict = None
    ) -> List[Dict[str, Any]]:
        """
        兜底路径：直接从 form_history 表查询，不依赖 FormTemplate。
        当 FormTemplate 无记录时使用（兼容旧数据或异常情况）。
        
        查询策略：
          1. 从 form_instances.data JSON 中反推 form_code（字段签名匹配）
          2. 找到匹配的 instance_ids
          3. 用这些 IDs 过滤 form_history
          4. 统计频次 + 时间衰减打分
        """
        from collections import defaultdict as ddict
        
        # 已知的字段签名 → form_code 映射
        FIELD_SIGNATURES = {
            "sales_order": {"customer_name", "customer_phone", "order_amount", "order_date", "remark"},
            "expense":    {"expense_type", "amount", "expense_date", "description"},
            "leave":      {"applicant_name", "department", "leave_type", "start_date", "end_date", "reason"},
        }
        
        # 1. 反推匹配的 instance_ids
        target_signature = FIELD_SIGNATURES.get(form_code)
        if not target_signature:
            logger.warning("[HistoryService] form_code=%s 无已知字段签名，跳过兜底查询", form_code)
            return []
        
        matched_instance_ids: List[int] = []
        instances = db.query(FormInstance.id, FormInstance.data).filter(
            FormInstance.status == 'submitted'
        ).all()
        
        for inst_id, data in instances:
            keys = set((data or {}).keys())
            overlap = len(keys & target_signature) / max(len(keys | target_signature), 1)
            if overlap >= 0.5:  # Jaccard >= 0.5 认为是同类型
                matched_instance_ids.append(inst_id)
        
        if not matched_instance_ids:
            return []
        
        # 2. 从 form_history 查询这些实例的字段值
        value_scores = ddict(lambda: {'count': 0, 'last_used': None})
        
        query_limit = config.get('historyQueryLimit', 1000) if config else 1000
        
        histories = db.query(FormHistory).filter(
            FormHistory.form_instance_id.in_(matched_instance_ids),
            FormHistory.field_code == field_code
        ).limit(query_limit).all()
        
        for h in histories:
            value = h.field_value or ''
            if value.strip():
                info = value_scores[value]
                info['count'] += 1
                if h.created_at and (not info['last_used'] or h.created_at > info['last_used']):
                    info['last_used'] = h.created_at
                if user_id and h.user_id == user_id:
                    info.setdefault('user_count', 0)
                    info['user_count'] += 1
        
        # 3. 打分排序
        recommendations = []
        now = datetime.now()
        
        for value, info in value_scores.items():
            score = cls._calculate_score(info, now, config)
            recommendations.append({
                'value': value,
                'score': score,
                'source': 'history_fallback',
                'count': info['count'],
                'user_count': info.get('user_count', 0)
            })
        
        rec_limit = config.get('recommendationLimit', 10) if config else 10
        result = sorted(recommendations, key=lambda x: x['score'], reverse=True)[:rec_limit]
        logger.info("[HistoryService] form_code=%s field=%s 兜底查询到 %d 条推荐",
                     form_code, field_code, len(result))
        return result

    @classmethod
    def save_history(
        cls, 
        form_instance_id: int, 
        field_code: str, 
        field_value: str, 
        user_id: str = None,
        db: Session = None
    ):
        if not db:
            return
        
        try:
            history = FormHistory(
                form_instance_id=form_instance_id,
                field_code=field_code,
                field_value=str(field_value) if field_value is not None else '',
                user_id=user_id
            )
            db.add(history)
            db.commit()
            logger.debug("[HistoryService] 保存历史记录 instance_id=%s field=%s",
                         form_instance_id, field_code)
        except Exception as e:
            logger.exception("[HistoryService] 保存历史记录失败 instance_id=%s field=%s: %s",
                             form_instance_id, field_code, e)
            db.rollback()

File: backend-app/src/main/java/com/sitech/prodai/service/HistoryService.java
package com.sitech.prodai.service;

import com.sitech.prodai.config.ConfigLoader;
import com.sitech.prodai.domain.entity.OntologyInstance;
import com.sitech.prodai.domain.entity.OntologyInstanceHistory;
import com.sitech.prodai.repository.OntologyInstanceHistoryRepository;
import com.sitech.prodai.repository.OntologyInstanceRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * 历史数据服务 —— 对齐 Python {@code app/services/history_service.py::HistoryService}。
 *
 * <p>提供基于历史数据的字段推荐（频次 + 时间衰减 + 用户个性化打分），
 * 同时承担 ontology_instance_history 表的写入。
 *
 * <p>核心算法：
 * <ul>
 *   <li>count_score = min(count * countScorePerUnit, 1.0) * countScoreWeight</li>
 *   <li>user_score = min(user_count * userScorePerUnit, 1.0) * userScoreWeight</li>
 *   <li>time_score = max(0, 1 - days_since / timeDecayDays) * timeScoreWeight</li>
 *   <li>final = count_score + user_score + time_score</li>
 * </ul>
 */
@Service
public class HistoryService {

    private static final Logger log = LoggerFactory.getLogger(HistoryService.class);

    private final ConfigLoader configLoader;
    private final OntologyService ontologyService;
    private final OntologyInstanceRepository instanceRepository;
    private final OntologyInstanceHistoryRepository historyRepository;

    public HistoryService(ConfigLoader configLoader,
                          OntologyService ontologyService,
                          OntologyInstanceRepository instanceRepository,
                          OntologyInstanceHistoryRepository historyRepository) {
        this.configLoader = configLoader;
        this.ontologyService = ontologyService;
        this.instanceRepository = instanceRepository;
        this.historyRepository = historyRepository;
    }

    // ==================== 推荐查询 ====================

    /** 对齐 Python get_recommend_values */
    public Map<String, Object> getRecommendValues(String formCode, String fieldCode,
                                                   String userId, Integer limit) {
        try {
            List<Map<String, Object>> recommendations = new ArrayList<>();
            Map<String, Object> config = configLoader.getRecommendationConfig();
            int limitValue = limit == null ? toInt(config.get("recommendationLimit"), 10) : limit;

            // 1. 数据库历史推荐
            List<Map<String, Object>> dbRecs = getDbRecommendations(formCode, fieldCode, userId, config);
            recommendations.addAll(dbRecs);

            // 2. 静态推荐兜底
            List<String> staticValues = configLoader.getRecommendations(formCode, fieldCode);
            Set<String> existingValues = new HashSet<>();
            for (Map<String, Object> r : recommendations) {
                existingValues.add(str(r.get("value")));
            }
            for (String rec : staticValues) {
                if (!existingValues.contains(rec)) {
                    Map<String, Object> r = new LinkedHashMap<>();
                    r.put("value", rec);
                    r.put("score", 0.5);
                    r.put("source", "static");
                    recommendations.add(r);
                    existingValues.add(rec);
                }
            }

            // 3. 排序
            recommendations.sort(Comparator.comparingDouble((Map<String, Object> r) ->
                    toDouble(r.get("score"))).reversed());

            // 4. 取 top N 并补充 label
            List<Map<String, Object>> top = new ArrayList<>();
            for (int i = 0; i < Math.min(limitValue, recommendations.size()); i++) {
                Map<String, Object> r = recommendations.get(i);
                String code = str(r.get("value"));
                String label = getFieldLabel(formCode, fieldCode, code);
                Map<String, Object> row = new LinkedHashMap<>();
                row.put("value", code);
                row.put("label", label);
                row.put("score", r.get("score"));
                row.put("source", r.getOrDefault("source", "db"));
                top.add(row);
            }

            Map<String, Object> body = new LinkedHashMap<>();
            body.put("success", true);
            body.put("recommendations", top);
            body.put("source", "hybrid");
            return body;
        } catch (Exception e) {
            log.error("[HistoryService] get_recommend_values 失败 form={} field={}", formCode, fieldCode, e);
            Map<String, Object> body = new LinkedHashMap<>();
            body.put("success", false);
            body.put("message", str(e));
            return body;
        }
    }

    // ==================== 写入 ====================

    /** 对齐 Python save_history */
    @Transactional
    public void saveHistory(Integer formInstanceId, String fieldCode, String fieldValue, String userId) {
        if (formInstanceId == null) {
            return;
        }
        try {
            OntologyInstanceHistory h = new OntologyInstanceHistory();
            h.setFormInstanceId(formInstanceId);
            h.setFieldCode(fieldCode);
            h.setFieldValue(fieldValue == null ? "" : fieldValue);
            h.setUserId(userId);
            historyRepository.save(h);
            log.debug("[HistoryService] 保存历史记录 instance_id={} field={}", formInstanceId, fieldCode);
        } catch (Exception e) {
            log.error("[HistoryService] 保存历史记录失败 instance_id={} field={}",
                    formInstanceId, fieldCode, e);
        }
    }

    // ==================== 内部方法 ====================

    /** 对齐 Python _get_db_recommendations */
    private List<Map<String, Object>> getDbRecommendations(String formCode, String fieldCode,
                                                            String userId, Map<String, Object> config) {
        if (config == null) {
            config = configLoader.getRecommendationConfig();
        }
        int queryLimit = toInt(config.get("historyQueryLimit"), 1000);
        int recLimit = toInt(config.get("recommendationLimit"), 10);

        // 查询 submitted 实例
        List<OntologyInstance> instances;
        if (userId != null && !userId.isEmpty()) {
            instances = instanceRepository.findByOntologyCodeAndStatusAndUserIdOrderBySubmittedAtDesc(
                    formCode, "submitted", userId, PageRequest.of(0, queryLimit));
        } else {
            instances = instanceRepository.findByOntologyCodeAndStatusOrderBySubmittedAtDesc(
                    formCode, "submitted", PageRequest.of(0, queryLimit));
        }

        // 统计 value → score_info
        Map<String, ScoreInfo> valueScores = new HashMap<>();
        for (OntologyInstance inst : instances) {
            Map<String, Object> data = inst.getData();
            if (data == null || !data.containsKey(fieldCode)) {
                continue;
            }
            Object valueObj = data.get(fieldCode);
            String value = valueObj == null ? "" : String.valueOf(valueObj).trim();
            if (value.isEmpty()) {
                continue;
            }
            ScoreInfo info = valueScores.computeIfAbsent(value, k -> new ScoreInfo());
            info.count++;
            LocalDateTime submitted = inst.getSubmittedAt();
            if (submitted != null) {
                if (info.lastUsed == null || submitted.isAfter(info.lastUsed)) {
                    info.lastUsed = submitted;
                }
            }
            if (userId != null && userId.equals(inst.getUserId())) {
                info.userCount++;
            }
        }

        LocalDateTime now = LocalDateTime.now();
        List<Map<String, Object>> recommendations = new ArrayList<>();
        for (Map.Entry<String, ScoreInfo> e : valueScores.entrySet()) {
            double score = calculateScore(e.getValue(), now, config);
            Map<String, Object> r = new LinkedHashMap<>();
            r.put("value", e.getKey());
            r.put("score", score);
            r.put("source", "database");
            r.put("count", e.getValue().count);
            r.put("user_count", e.getValue().userCount);
            recommendations.add(r);
        }
        recommendations.sort(Comparator.comparingDouble((Map<String, Object> r) ->
                toDouble(r.get("score"))).reversed());
        return recommendations.size() > recLimit ? recommendations.subList(0, recLimit) : recommendations;
    }

    /** 对齐 Python _calculate_score */
    private double calculateScore(ScoreInfo info, LocalDateTime now, Map<String, Object> config) {
        double countScoreWeight = toDouble(config.get("countScoreWeight"), 0.4);
        double userScoreWeight = toDouble(config.get("userScoreWeight"), 0.4);
        double timeScoreWeight = toDouble(config.get("timeScoreWeight"), 0.2);
        double countScorePerUnit = toDouble(config.get("countScorePerUnit"), 0.1);
        double userScorePerUnit = toDouble(config.get("userScorePerUnit"), 0.2);
        double timeDecayDays = toDouble(config.get("timeDecayDays"), 30);

        double countScore = Math.min(info.count * countScorePerUnit, 1.0);
        double userScore = Math.min(info.userCount * userScorePerUnit, 1.0);
        double timeScore = 1.0;
        if (info.lastUsed != null) {
            long daysSince = ChronoUnit.DAYS.between(info.lastUsed, now);
            if (daysSince > 0) {
                timeScore = Math.max(0.0, 1.0 - (daysSince / timeDecayDays));
            }
        }
        return countScore * countScoreWeight + userScore * userScoreWeight + timeScore * timeScoreWeight;
    }

    /** 对齐 Python _get_field_label —— 从本体定义查中文标签 */
    @SuppressWarnings("unchecked")
    private String getFieldLabel(String formCode, String fieldCode, String code) {
        try {
            Map<String, Object> ontologyResult = ontologyService.getFormConstraint(formCode);
            if (!Boolean.TRUE.equals(ontologyResult.get("success"))) {
                return code;
            }
            Object constraintsObj = ontologyResult.get("constraints");
            if (!(constraintsObj instanceof Map<?, ?> constraints)) {
                return code;
            }
            Object entitiesObj = constraints.get("entities");
            if (!(entitiesObj instanceof List<?> entities)) {
                return code;
            }
            for (Object entityObj : entities) {
                if (!(entityObj instanceof Map<?, ?> entity)) {
                    continue;
                }
                Object fieldsObj = entity.get("fields");
                if (!(fieldsObj instanceof List<?> fields)) {
                    continue;
                }
                for (Object fieldObj : fields) {
                    if (!(fieldObj instanceof Map<?, ?> field)) {
                        continue;
                    }
                    if (!fieldCode.equals(field.get("fieldCode"))) {
                        continue;
                    }
                    // 优先从 enumConfig.options 查找
                    List<?> options = List.of();
                    Object enumConfigObj = field.get("enumConfig");
                    if (enumConfigObj instanceof Map<?, ?> enumConfig) {
                        Object opts = enumConfig.get("options");
                        if (opts instanceof List<?> list) {
                            options = list;
                        }
                    }
                    if (options.isEmpty()) {
                        Object opts = field.get("options");
                        if (opts instanceof List<?> list) {
                            options = list;
                        }
                    }
                    for (Object opt : options) {
                        if (opt instanceof String s) {
                            if (s.equals(code)) {
                                return s;
                            }
                        } else if (opt instanceof Map<?, ?> optMap) {
                            if (code.equals(optMap.get("value"))) {
                                Object label = optMap.get("label");
                                return label == null ? code : String.valueOf(label);
                            }
                        }
                    }
                    return code;
                }
            }
            return code;
        } catch (Exception e) {
            log.warn("[HistoryService] 获取字段标签失败: {}", e.getMessage());
            return code;
        }
    }

    // ==================== 工具方法 ====================

    private int toInt(Object value, int defaultValue) {
        if (value instanceof Number n) {
            return n.intValue();
        }
        if (value == null) {
            return defaultValue;
        }
        try {
            return Integer.parseInt(String.valueOf(value));
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    private double toDouble(Object value) {
        return toDouble(value, 0.0);
    }

    private double toDouble(Object value, double defaultValue) {
        if (value instanceof Number n) {
            return n.doubleValue();
        }
        if (value == null) {
            return defaultValue;
        }
        try {
            return Double.parseDouble(String.valueOf(value));
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    private String str(Object value) {
        return value == null ? "" : String.valueOf(value);
    }

    private String str(Throwable e) {
        return e.getMessage() == null ? e.getClass().getSimpleName() : e.getMessage();
    }

    /** 内部打分信息 */
    private static class ScoreInfo {
        int count;
        int userCount;
        LocalDateTime lastUsed;
    }
}
